tls/extensions/s2n_client_key_share.c

```diff
@@ -56,30 +59,57 @@ const s2n_extension_type s2n_client_key_share_extension = {
     .if_missing = s2n_extension_noop_if_missing,
 };
 
-static int s2n_ecdhe_supported_curves_send(struct s2n_connection *conn, struct s2n_stuffer *out)
+static int s2n_generate_preferred_key_shares(struct s2n_connection *conn, struct s2n_stuffer *out)
 {
-    notnull_check(conn);
-    notnull_check(conn->config);
+    uint8_t preferred_key_shares = conn->preferred_key_shares;
+    const struct s2n_ecc_named_curve *named_curve = NULL;
+    struct s2n_ecc_evp_params *ecc_evp_params = NULL;
 
     const struct s2n_ecc_preferences *ecc_pref = NULL;
     GUARD(s2n_connection_get_ecc_preferences(conn, &ecc_pref));
     notnull_check(ecc_pref);
 
-    const struct s2n_ecc_named_curve *named_curve = NULL;
-    struct s2n_ecc_evp_params *ecc_evp_params = NULL;
+    bool empty_keyshares = preferred_key_shares & 1;
 
-    for (uint32_t i = 0; i < ecc_pref->count; i++) {
-        ecc_evp_params = &conn->secure.client_ecc_evp_params[i];
-        named_curve = ecc_pref->ecc_curves[i];
+    if (!conn->preferred_key_shares) {
+        /* Default behavior is to generate keyshares for all curves.
+        * The bitmap to generate keyshares for all curve is 111111110 (254),
+        * i.e. all bit values set except lsb which is RESERVED for empty keyshares */
+        preferred_key_shares = GENERATE_KEYSHARES_ALL_CURVES;
+    }
 
+    for (size_t i = 1; i <= ecc_pref->count; i++) {
+        ecc_evp_params = &conn->secure.client_ecc_evp_params[i-1];
+        named_curve = ecc_pref->ecc_curves[i-1];
         ecc_evp_params->negotiated_curve = named_curve;
         ecc_evp_params->evp_pkey = NULL;
-        GUARD(s2n_ecdhe_parameters_send(ecc_evp_params, out));
+        /* If lsb is set, skip keyshare generation for all curve */
+        if (empty_keyshares) {
+            GUARD(s2n_stuffer_write_uint16(out, ecc_evp_params->negotiated_curve->iana_id));
```

Comment:
As is, this is writing a vector of zeroed out keyshares. Per https://tools.ietf.org/html/rfc8446#section-4.2.8 we instead need to write an empty vector(write 0 for `client_shares` length) when no key shares are provided.

Verified a 1.3 server does not like how the empty key shares are currently encoded:
```
> ./bin/s2nc -i -k '0' --tls13 --ciphers 'default_tls13' www.cloudflare.com 443
Failed to negotiate: 'TLS alert received'. Error encountered in s2n_alerts.c line 97
Alert: 47
> ./bin/s2nc -i --tls13 --ciphers 'default_tls13' www.cloudflare.com 443
CONNECTED:
Client hello version: 33
Client protocol version: 34
Server protocol version: 34
Actual protocol version: 34
Server name: www.cloudflare.com
Curve: x25519
KEM: NONE
Cipher negotiated: TLS_AES_256_GCM_SHA384
Connected to www.cloudflare.com:443
```

Reply:
@raycoll The reason I went the route of writing a vector of zeroed out keyshares instead of an empty keyshare with NULL is because for x25519, a NULL keyshare is considered a valid keyshare. Ref: https://moderncrypto.org/mail-archive/curves/2017/000896.html.

Reply:
https://moderncrypto.org/mail-archive/curves/2017/000896.html is interesting but I don't think it's related to this problem(how to write 0 key shares as a TLS client).

* What this change is writing to signal no key shares(and causing handshake failures against TLS 1.3 servers):

![image](https://user-images.githubusercontent.com/4444641/83454738-7f939880-a411-11ea-9db6-6c96273fade4.png)

<img width="589" alt="image" src="https://user-images.githubusercontent.com/4444641/83454257-ca60e080-a410-11ea-8b21-482b4deeaa83.png">

* What should be written:

<img width="391" alt="image" src="https://user-images.githubusercontent.com/4444641/83454647-5bd05280-a411-11ea-87c5-55e91d15c864.png">

![image](https://user-images.githubusercontent.com/4444641/83454432-085e0480-a411-11ea-8710-c8644ebcdc5f.png)



The `else` block of this conditional also has the same issue(writing an entry rather than just skipping the group entirely). 

(pardon the screenshots, I couldn't figure out a clear way to write ASN1 syntax using text)